v2 api with gRPC and gRPC gateway

KeyboardNerd opened the pull request:
It's a big commit, probably the logging should be improved. ( currently it's a hack on response writer )

jzelinskie commented on api/api.go:
```diff
@@ -68,12 +81,12 @@ func Run(cfg *Config, store database.Datastore, st *stopper.Stopper) {
 	log.Info("main API stopped")
 }
 
+// RunHealth API
```
this isn't a particularly useful comment

jzelinskie commented on api/api.go:
```diff
@@ -35,15 +36,27 @@ const timeoutResponse = `{"Error":{"Message":"Clair failed to respond within the
+	v2.Run(cfg.GrpcPort, tlsConfig, cfg.PaginationKey, cfg.CertFile, cfg.KeyFile, store)
+}
+
+// Run V1 API
```
this isn't a particularly useful comment

jzelinskie commented on api/v1/routes.go:
```diff
@@ -24,7 +24,7 @@ import (
 	"github.com/julienschmidt/httprouter"
 	"github.com/prometheus/client_golang/prometheus"
-	log "github.com/sirupsen/logrus"
+	log "github.com/Sirupsen/logrus"
```
In some files this is the first import and in this one it's the last. Did you run GoImports on this? They should be ordered alphabetically.

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -0,0 +1,133 @@
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+
+message Vulnerability{
```
Is this file in an idiomatic format? The non-space-separated `{` seems strange.
There's got to be a protobuf linter that has opinions on this stuff.

KeyboardNerd replied:
Nope; based on the style, I shouldn't use the camel case for fields: https://developers.google.com/protocol-buffers/docs/style

jzelinskie replied:
Change this so that you always have the space in between the message name and the curly bracket `message Name {` instead of `message Name{`

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -0,0 +1,133 @@
+
+message VulnerabilityWithLayers{
+	Vulnerability Vulnerability = 1;
+	repeated OrderedLayerName OrderedLayersIntroducingVulnerability = 2;
```
We can just called this `LayersIntroducingVulnerabilty`, it was given this verbose name because the original `LayersIntroducingVulnerability` was unordered and we didn't want to break backwards-compat in the v1 API.

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -0,0 +1,133 @@
+	string Name = 1;
+}
+
+message ResPostAncestry{
```
Let's use the naming convention: `GetNotificationRequest` and `PostAncestryResponse`. This is what etcd does.

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+package v2
```
This file is missing the license header

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	"github.com/coreos/clair/pkg/commonerr"
+)
+
+// ClairAPIServer is the gRPC server for v2 API
```
Doc comments need to be full sentences ending with a period wrapped at 80char lines.

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+// PostAncestry is used by the gRPC server for handling Post ancestry request
+func (s *ClairAPIServer) PostAncestry(c context.Context, r *pb.ReqPostAncestry) (*pb.ResPostAncestry, error) {
```
it's idiomatic to use `ctx` as the name for a context.
`req` is also probably a better name than `r`
these need to be done to the whole file and not just this instance.

KeyboardNerd committed:
api: v2 api with gRPC and gRPC-gateway

Newly designed API defines Ancestry as a set of layers
and shrinked the api to only the most used apis:
post ancestry, get layer, get notification, delete notification

Fixes #98

KeyboardNerd committed:
vendor: updated vendor dir for grpc v2 api

jzelinskie commented on .travis.yml:
```diff
@@ -1,4 +1,8 @@
 language: go
+
+go: 
+  - 1.8.3
```
what was it using before?
this should probably be just "1.8" so that we get tested on future patches

KeyboardNerd replied:
travis was using go1.7.4 linux/amd64 ( default behavior? )

jzelinskie replied:
yeah, this is fine, if we just change it to 1.8

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+			return nil, s.rollBackOnError(err, currentName, rootName)
+		}
+
+		// TODO: after altering the database to support ancestry, we should use the ancestry name and index as key instead of
```
comments wrapped at 80chars
`TODO(keyboardnerd): ...`

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+			// root name is the name of root layer in an ancestry
+			rootName = currentName
+		}
+		err := clair.ProcessLayer(s.Store, ancestry.GetFormat(), currentName, parentName, layer.GetPath(), layer.GetHeaders())
```
newline between this and the rootName check

code blocks should all be logically connected

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+		}
+
+		if rootName == "" {
+			// root name is the name of root layer in an ancestry
```
remove this comment and put another above the `if` that says something along the lines of:

> If rootName is unset, this is the first iteration over the layers and the current layer is the root of the ancestry.

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+		if err != nil {
+			return nil, s.rollBackOnError(err, currentName, rootName)
+		}
+		// every layer's parent should be on its index - 1 and child should be on its index + 1
```
Let's rewrote this comment to be similar to the one before.

>Now that the current layer is processed, set the parentName for the next iteration.

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	}
+	res.Ancestry = ancestry
+	res.EngineVersion = clair.Version
+	return res, nil
```
remove L37 where you do `res := &pb.ResPostAncestry{}`
and replace L67-69 with:

```go
return &pb.ResPostAncestry{
  Ancestry: ancestry,
  EngineVersion: clair.Version,
}, nil
```

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	if r.GetAncestryName() == "" || r.GetIndex() < 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid get layer request")
+	}
+	// TODO: after altering the database to support ancestry, this function is iteratively querying for for r.GetIndex() th parent of the
```
comments wrapped at 80 chars
TODO(keyboardnerd): ...

jzelinskie replied:
`TODO(keyboardnerd):` 80 char lines full sentences with capitalization etc...

jzelinskie commented on api/v2/util.go:
```diff
@@ -0,0 +1,250 @@
+	return &vuln, nil
+}
+
+func layerFromDatabaseModel(dbLayer database.Layer, withFeatures, withVulnerabilities bool) (*pb.Layer, error) {
```
Methods like `XFromDatabaseModel` could be exported and put in the same package as the generated protobufs.

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+	return &pb.ResGetAncestry{Layer: layer, EngineVersion: int32(dbLayer.EngineVersion)}, nil
```
multi-line this please

```go
return &pb.ResGetAncestry{
  Layer: layer,
  EngineVersion: int32(dbLayer.EngineVersion)
}, nil
```

jzelinskie replied:
2+ fields break into newlines

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+
+// GetNotification returns a notification response
+func (s *ClairAPIServer) GetNotification(c context.Context, r *pb.ReqGetNotification) (*pb.ResGetNotification, error) {
+	// validate
```
remove this comment

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	return &pb.ResGetAncestry{Layer: layer, EngineVersion: int32(dbLayer.EngineVersion)}, nil
+}
+
+// GetNotification returns a notification response
```
let's rewrite all the comments in this form into

> // GetNotification implements retrieving a notification via the Clair gRPC service.
>// GetAncestry implements retrieving an ancestry via the Clair gRPC service.

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	if r.GetLimit() <= 0 {
+		return nil, status.Error(codes.InvalidArgument, "Failed to provide page limit")
+	}
+	// get page token
```
just replace this comment with a new line

jzelinskie replied:
remove these comments

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+		pageToken = string(pageTokenBytes)
+	}
+
+	// get the notification
```
remove this comment

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -0,0 +1,146 @@
+	} else if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &google_protobuf1.Empty{}, nil
```
newline above this

jzelinskie replied:
Is this idiomatic? I'm going to ask someone.

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+package v2
```
license header

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+		grpcOpts = append(grpcOpts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+	}
+	grpcServer := grpc.NewServer(grpcOpts...)
+	pb.RegisterClairAPIServiceServer(grpcServer, &ClairAPIServer{PaginationKey: paginationKey, Store: store})
```
struct literals with 2+ fields you should put into multiple lines

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+		if lrw.StatusCode == 0 {
+			statusStr = "???"
+		}
+		log.WithFields(log.Fields{"remote addr": r.RemoteAddr, "method": r.Method, "request uri": r.RequestURI, "status": statusStr, "elapsed time": time.Since(start)}).Info("Handled HTTP request")
```
multiple lines, please

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+		tlsConfig.NextProtos = []string{"h2"} // enable http2 support
+		// setup tls listener
```
remove these comments, they aren't useful

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+	"github.com/coreos/clair/database"
+)
+
+func errorHandle(err error) {
```
this function needs a better name and a doc comment describing what it does

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+		servePrometheus(http1Mux)
+		handler := grpcHandlerFunc(grpcServer, http1Mux)
+
+		srv := &http.Server{
```
we can declare this before the `if tlsConfig != nil` and then have one `errorHandle(srv.Serve(listener))` at the bottom of the function to make it clear what's blocking when reading the function

there might be a couple more things we can pull out to try and keep the conditional logic smaller

jzelinskie commented on api/v2/util.go:
```diff
@@ -0,0 +1,250 @@
+package v2
```
license

jzelinskie commented on api/v2/util.go:
```diff
@@ -0,0 +1,250 @@
+	"fmt"
+	"time"
+
+	fernet "github.com/fernet/fernet-go"
```
i don't think you need to qualify this import, it should already be imported as "fernet"

jzelinskie commented on api/v2/server.go:
```diff
@@ -0,0 +1,183 @@
+		statusStr := strconv.Itoa(lrw.StatusCode)
+		if lrw.StatusCode == 0 {
+			statusStr = "???"
+		}
```
We should probably stop doing this and and just assume all Prometheus requests responded with 200, but that can be handled in a separate PR.

jzelinskie commented on api/v2/util.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+// TODO: Remove this Legacy compability code once the database is revised
+// We assume every layer is the root of an ancestry for api being compatible with the current implementation of the database
```
comment rules

jzelinskie commented on api/v2/util.go:
```diff
@@ -0,0 +1,250 @@
+		layer database.Layer
+		err   error
+	)
+	// iteratively query the layer until we find the one layer requested
```
comment rules

jzelinskie commented on api/token/token.go:
```diff
@@ -0,0 +1,48 @@
+	"errors"
+	"time"
+
+	fernet "github.com/fernet/fernet-go"
```
Do you actually need this? I'm pretty sure without it, it still imports as `fernet`.

KeyboardNerd replied:
forgot to check, it's automatically imported like that

jzelinskie commented on api/token/token.go:
```diff
@@ -0,0 +1,48 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package token
```
// Package token implements ...

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -0,0 +1,147 @@
+
+message LayersIntroducingVulnerabilty{
+	Vulnerability vulnerability = 1;
+	repeated OrderedLayerName ordered_layers_introducing_vulnerability = 2;
```
`layers_introducing_vulnerability` or maybe even just `layers` is sufficient because of the message's name and `OrderedLayerName` tells you that it's ordered

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -0,0 +1,147 @@
+message Layer{
+	string name = 1;
+	repeated string namespace_names = 2;
+	string parent_name = 3;
```
Do we need this in v2?

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -0,0 +1,147 @@
+	Notification notification = 1;
+}
+
+service ClairAPIService{
```
I talked to the etcd team and I think it's a nice pattern to separate Services by resource, so we can have two Services that we register to the same grpc server.

```protobuf
service AncestryService {
   rpc GetAncestry(GetAncestryRequest) returns (GetAncestryResponse) { ... };
   rpc PostAncestry(PostAncestryRequest) returns (PostAncestryResponse) { ... };
}

service NotificationService {
  rpc GetNotification(GetNotificationRequest) returns (GetNotificationResponse) { ... };
  rpc DeleteNotification(DeleteNotificationRequest) returns (DeleteNotificationResponse) { ... };
}
```
